test/extended/util/disruption/backend_sampler_tester.go

```diff
@@ -0,0 +1,124 @@
+	"k8s.io/kubernetes/test/e2e/upgrades"
+)
+
+type BackendSampler interface {
+	GetDisruptionBackendName() string
+	GetURL() (string, error)
+	SetEventRecorder(recorder events.EventRecorder)
+	StartEndpointMonitoring(ctx context.Context, m *monitor.Monitor) error
+}
+
+func NewBackendDisruptionTest(testName string, backend BackendSampler) *backendDisruptionTest {
+	return &backendDisruptionTest{
+		testName:             testName,
+		backend:              backend,
+		getAllowedDisruption: NoDisruption,
+	}
+}
+
+func (t *backendDisruptionTest) WithAllowedDisruption(allowedDisruptionFn AllowedDisruptionFunc) *backendDisruptionTest {
+	t.getAllowedDisruption = allowedDisruptionFn
+	return t
+}
+
+type SetupFunc func(f *framework.Framework) error
+
+func (t *backendDisruptionTest) WithPreSetup(preSetup SetupFunc) *backendDisruptionTest {
+	t.preSetup = preSetup
+	return t
+}
+
+type TearDownFunc func(f *framework.Framework) error
+
+func (t *backendDisruptionTest) WithPostTeardown(postTearDown TearDownFunc) *backendDisruptionTest {
+	t.postTearDown = postTearDown
+	return t
+}
+
+func NoDisruption(f *framework.Framework, totalDuration time.Duration) (*time.Duration, error) {
+	zero := 0 * time.Second
+	return &zero, nil
+}
+
+type AllowedDisruptionFunc func(f *framework.Framework, totalDuration time.Duration) (*time.Duration, error)
+
+// availableTest tests that route frontends are available before, during, and
+// after a cluster upgrade.
+type backendDisruptionTest struct {
+	// testName is the name to show in unit.
+	testName string
+	// backend describes a route that should be monitored.
+	backend              BackendSampler
+	getAllowedDisruption AllowedDisruptionFunc
+
+	preSetup     SetupFunc
+	postTearDown TearDownFunc
+}
+
+func (t *backendDisruptionTest) Name() string { return t.backend.GetDisruptionBackendName() }
+func (t *backendDisruptionTest) DisplayName() string {
+	return t.testName
+}
+
+// Setup looks up the host of the route specified by the backendSampler and updates
+// the backendSampler with the route's host.
+func (t *backendDisruptionTest) Setup(f *framework.Framework) {
+	if t.preSetup != nil {
+		framework.ExpectNoError(t.preSetup(f))
+	}
+
+	url, err := t.backend.GetURL()
+	framework.ExpectNoError(err)
+	if len(url) == 0 {
+		framework.Failf("backend has no URL: %#v", t.backend)
+	}
+}
+
+// Test runs a connectivity check to a route.
+func (t *backendDisruptionTest) Test(f *framework.Framework, done <-chan struct{}, upgrade upgrades.UpgradeType) {
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	newBroadcaster := events.NewBroadcaster(&events.EventSinkImpl{Interface: f.ClientSet.EventsV1()})
+	t.backend.SetEventRecorder(newBroadcaster.NewRecorder(scheme.Scheme, "openshift.io/"+t.Name()))
+	newBroadcaster.StartRecordingToSink(stopCh)
+
+	ginkgo.By("continuously hitting backend")
+
+	ctx, cancel := context.WithCancel(context.Background())
+	m := monitor.NewMonitorWithInterval(1 * time.Second)
```

Comment:
hm...... This may end up causing us to double count apiserver outages because two different monitors will be failing.  What we would *like* have to happen is that the interval calculation correctly handles, "we were failing for an overlapping period of time".